distributed/scheduler.py

```diff
@@ -445,16 +445,23 @@ def close_worker(self, stream=None, worker=None):
         """
         with log_errors():
             original = worker
-            self.remove_worker(address=original)
-            with ignoring(KeyError):
+            try:
                 nanny_port = self.worker_info[worker]['services']['nanny']
+            except KeyError:
+                nanny_port = False
+
+            self.remove_worker(address=worker)
+
+            if nanny_port:
                 ip, port = worker.split(':')
-                worker = '%s:%s' % (ip, nanny_port)
+                address = '%s:%s' % (ip, nanny_port)
+            else:
+                address = worker
 
-            with rpc(worker) as r:
+            with rpc(addr=address) as r:
                 yield r.terminate(report=False)
 
-            self.remove_worker(address=original)
+            self.remove_worker(address=worker)
```

Comment:
remove_worker is idempotent.  There is a slight chance that we have received an out-of-order heartbeat (which is just an add-worker message) during the yield.  I didn't actually experience this, nor do I have confidence that it would be an issue, but it is cheap and safe to call remove_worker the second time.